### PR TITLE
[Flang] [OpenMP] [Semantics] [MLIR] [Lowering] Add lowering support for IS_DEVICE_PTR and HAS_DEVICE_ADDR clauses on OMP TARGET directive.

### PR DESCRIPTION
**Pull request by raghavendhra**
Added lowering support for IS_DEVICE_PTR and HAS_DEVICE_ADDR clauses for OMP TARGET directive and added related tests for these changes.

IS_DEVICE_PTR and HAS_DEVICE_ADDR clauses apply to OMP TARGET directive OpenMP spec states

`The **is_device_ptr** clause indicates that its list items are device pointers.`

`The **has_device_addr** clause indicates that its list items already have device addresses and therefore they may be directly accessed from a target device.`

Whereas USE_DEVICE_PTR and USE_DEVICE_ADDR clauses apply to OMP TARGET DATA directive and OpenMP spec for them states

`Each list item in the **use_device_ptr** clause results in a new list item that is a device pointer that refers to a device address`

`Each list item in a **use_device_addr** clause that is present in the device data environment is treated as if it is implicitly mapped by a map clause on the construct with a map-type of alloc`

### CODE REVIEW

**Comment by raghavendhra** [comment]
> It would be nice to split semantic from lowering changes.

Created a https://github.com/llvm/llvm-project/pull/74370 to address semantic part of this PR. Once https://github.com/llvm/llvm-project/pull/74370 is accepted I will rebase this patch.

**Comment by raghavendhra** [comment]
> > It would be nice to split semantic from lowering changes.
> 
> Created a #74370 to address semantic part of this PR. Once #74370 is accepted I will rebase this patch.

https://github.com/llvm/llvm-project/pull/74370 Has been accepted and merged to upstream. @clementval 
Can you please review this now?

**Review by kiranchandramohan — COMMENTED** [review]
Please add additional information to the description on how these is related to `use_device_addr` and `use_device_ptr`. Also add information on how this ties with the the `IsolatedFromAbove` property of the target region.

**Comment by raghavendhra** [comment]
Ping for review

**Review by skatrak — APPROVED** [review]
Thank you Raghu, this LGTM. I just have some small nits, and also I think the PR/commit title should be updated to remove the "Semantics" tag. Please wait for at least a day before merging, in case there are remaining concerns by other reviewers.

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td [hunk 1]
```diff
@@ -1628,14 +1628,23 @@ def TargetOp : OpenMP_Op<"target", [IsolatedFromAbove, MapClauseOwningOpInterfac
...
-    The optional $nowait elliminates the implicit barrier so the parent task can make progress
+    The optional $nowait eliminates the implicit barrier so the parent task can make progress
     even if the target task is not yet completed.
 
     The `depends` and `depend_vars` arguments are variadic lists of values
     that specify the dependencies of this particular target task in relation to
     other tasks.
 
-    TODO:  is_device_ptr, defaultmap, in_reduction
+    The optional $is_device_ptr indicates list items are device pointers
```
[the remark this answers is not in the document]
Comment (raghavendhra) [reply mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:1638]: Done!

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td [hunk 2]
```diff
@@ -1628,14 +1628,23 @@ def TargetOp : OpenMP_Op<"target", [IsolatedFromAbove, MapClauseOwningOpInterfac
...
+
+    The optional $has_device_addr indicates that list items already have device
+    addresses, so may be directly accessed from target device. May include array
+    sections.
```
[the remark this answers is not in the document]
Comment (raghavendhra) [reply mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:1642]: Done!

**Comment by jplehr** [comment]
Hi,
this broke two of our bots
https://lab.llvm.org/staging/#/builders/140/builds/5923
https://lab.llvm.org/buildbot/#/builders/285/builds/54

Let me know if you need assistance in reproducing the error.

**Comment by raghavendhra** [comment]
[AMD Official Use Only - General]

Hi JP,

Thank you for letting me know. I just started looking into it.

--Raghu
________________________________

**Comment by Dinistro** [comment]
Can confirm, this leads to a compilation error in `OpenMPDialect.cpp`.

**Pull request merged**